File: api/app/domain/models/plan.py
from typing import Optional
from typing import List
from typing import Any
import uuid
from pydantic import Field
from pydantic import BaseModel
from enum import Enum


class ExecutionStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class Step(BaseModel):
    """计划中的每一个步骤/子任务"""

    id: str = Field(default_factory=lambda: str(uuid.uuidv4()))
    description: str = ""  # 子任务描述信息
    status: ExecutionStatus = ExecutionStatus.PENDING
    result: Optional[str] = None  # 结果
    error: Optional[str] = None  # 错误信息
    success: bool = False  # 是否执行成功
    attachments: Optional[List[str]] = None  # 附件列表


class Plan(BaseModel):
    """规划Domain模型，存储用户传递消息拆分出来的子任务、子步骤"""

    id: str = Field(default_factory=lambda: str(uuid.uuidv4()))
    title: str = ""
    goal: str = ""
    language: str = ""
    message: str = ""
    status: ExecutionStatus = ExecutionStatus.PENDING
    error: Optional[str] = None
    steps: List[Any] = Field(default_factory=list)
    result: Optional[str] = None  # 最终结果

    @property
    def done(self) -> bool:
        return self.status in [ExecutionStatus.COMPLETED, ExecutionStatus.FAILED]

    def get_next_step(self) -> Optional[Step]:
        """获取下一个未执行的步骤"""
        return next((step for step in self.steps if not step.done), None)

File: api-node/src/domain/plan.ts
import { v4 as uuidv4 } from 'uuid';

export enum ExecutionStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  COMPLETED = 'completed',
  FAILED = 'failed',
}

export interface Step {
  id: string;
  description: string;
  status: ExecutionStatus;
  result?: string | null;
  error?: string | null;
  success: boolean;
  attachments?: string[] | null;
}

export function isStepDone(step: Step): boolean {
  return step.status === ExecutionStatus.COMPLETED || step.status === ExecutionStatus.FAILED;
}

export interface Plan {
  id: string;
  title: string;
  goal: string;
  language: string;
  message: string;
  status: ExecutionStatus;
  error?: string | null;
  steps: Step[];
  result?: string | null;
}

export function createStep(description: string, id = uuidv4()): Step {
  return {
    id,
    description,
    status: ExecutionStatus.PENDING,
    success: false,
    attachments: null,
  };
}

export function parsePlan(raw: Record<string, unknown>): Plan {
  const steps = Array.isArray(raw.steps)
    ? raw.steps.map((s: Record<string, unknown>) => ({
        id: String(s.id ?? uuidv4()),
        description: String(s.description ?? ''),
        status: ExecutionStatus.PENDING,
        success: false,
        result: null,
        error: null,
        attachments: null,
      }))
    : [];

  return {
    id: String(raw.id ?? uuidv4()),
    title: String(raw.title ?? ''),
    goal: String(raw.goal ?? ''),
    language: String(raw.language ?? 'Chinese'),
    message: String(raw.message ?? ''),
    status: ExecutionStatus.PENDING,
    steps,
    result: null,
    error: null,
  };
}

export function parseStepResult(raw: Record<string, unknown>): Partial<Step> {
  return {
    success: Boolean(raw.success),
    result: raw.result != null ? String(raw.result) : null,
    attachments: Array.isArray(raw.attachments) ? raw.attachments.map(String) : null,
  };
}

export function getNextStep(plan: Plan): Step | null {
  return plan.steps.find((s) => !isStepDone(s)) ?? null;
}
